src/translate.py

```diff
@@ -51,100 +51,45 @@
     TypePool,
 )
 
+InstrSet = Set[str]
+InstrMap = Dict[str, Callable[["InstrArgs"], "Expression"]]
+StmtInstrMap = Dict[str, Callable[["InstrArgs"], "Statement"]]
+CmpInstrMap = Dict[str, Callable[["InstrArgs"], "Condition"]]
+StoreInstrMap = Dict[str, Callable[["InstrArgs"], Optional["StoreStmt"]]]
+MaybeInstrMap = Dict[str, Callable[["InstrArgs"], Optional["Expression"]]]
+PairInstrMap = Dict[str, Callable[["InstrArgs"], Tuple["Expression", "Expression"]]]
+
+
+class Arch(abc.ABC):
+    stack_pointer_reg: Register
+    frame_pointer_reg: Register
+    return_address_reg: Register
+    temporary_reg: Register
```

Comment:
so, since `PpcArch` won't have `temporary_reg` (right?), are we going to do `Optional[Register]`? How do we know when we're "over-generifying" the situation?

Reply:
I decided to convert all `Register(...)` constants in `translate.py` to one of these class members, *except* in the following cases:
- `Register("zero")`, this is already a special value used by `parse_instruction.py`
- Any code related to function arguments, which will need to be replaced next anyway (and probably won't be replaced with a simple list).
- Usages of `Register("condition_bit")`, `Register("hi")`, `Register("lo")` as they relate to `Arch.instrs_float_comp` and `Arch.instrs_hi_lo`. These are used as the implicit destination(s) of certain instructions. I also figured these might get refactored when adding support for PPC, they might fit into the `CASES_IMPLICIT_DESTINATION` paradigm? 

On the other hand, I sort of expect to convert `frame_pointer_reg` and `temporary_reg` into `Optional[Register]` when adding PPC support. (But I chose to wait until it's explicitly needed?)
- `Register("ra")` is used in `overwrite_reg(...)` even if the instruction may not reference `$ra` directly. It's straightforward to convert this code to use an `Optional[Register]` instead.
- `Register("fp")` is related to `StackInfo.uses_framepointer`. I imagine for PPC we could keep `StackInfo.uses_framepointer = False` and `PpcArch.frame_pointer_reg = None`.

I guess it's a bit of a personal taste how much "preemptive refactoring" makes sense to do. I wanted to abstract away enough registers to make sure that I'd have access to the `arch` variable in all the places that need it.

Reply:
(Re overwrite_reg it would be equally fine to remove that code, it's already if False'd)